input/witnessgen.go

```diff
@@ -401,6 +510,205 @@ func (wt StandardWitnessType) WitnessGenerator(signer Signer,
 		case NestedWitnessKeyHash:
 			return signer.ComputeInputScript(tx, desc)
 
+		case TaprootLocalCommitSpend:
+			// Ensure that the sign desc has the proper sign method
+			// set, and a valid prev output fetcher.
+			desc.SignMethod = TaprootScriptSpendSignMethod
+
+			// The control block bytes must be set at this point.
+			if desc.ControlBlock == nil {
+				return nil, fmt.Errorf("control block must be " +
+					"set for taproot spend")
+			}
+
+			witness, err := TaprootCommitSpendSuccess(
+				signer, desc, tx, nil,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			return &Script{
+				Witness: witness,
+			}, nil
+
+		case TaprootRemoteCommitSpend:
+			// Ensure that the sign desc has the proper sign method
+			// set, and a valid prev output fetcher.
+			desc.SignMethod = TaprootScriptSpendSignMethod
+
+			// The control block bytes must be set at this point.
+			if desc.ControlBlock == nil {
+				return nil, fmt.Errorf("control block must be " +
+					"set for taproot spend")
+			}
+
+			witness, err := TaprootCommitRemoteSpend(
+				signer, desc, tx, nil,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			return &Script{
+				Witness: witness,
+			}, nil
+
+		case TaprootAnchorSweepSpend:
+			// Ensure that the sign desc has the proper sign method
+			// set, and a valid prev output fetcher.
+			desc.SignMethod = TaprootKeySpendSignMethod
+
+			// The tap tweak must be set at this point.
+			if desc.TapTweak == nil {
+				return nil, fmt.Errorf("tap tweak must be " +
+					"set for keyspend")
+			}
+
+			witness, err := TaprootAnchorSpend(
+				signer, desc, tx,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			return &Script{
+				Witness: witness,
+			}, nil
+
+		case TaprootHtlcOfferedTimeoutSecondLevel,
+			TaprootHtlcAcceptedSuccessSecondLevel:
+			// Ensure that the sign desc has the proper sign method
+			// set, and a valid prev output fetcher.
+			desc.SignMethod = TaprootScriptSpendSignMethod
+
+			// The control block bytes must be set at this point.
+			if desc.ControlBlock == nil {
+				return nil, fmt.Errorf("control block must be " +
+					"set for taproot spend")
+			}
+
+			witness, err := TaprootHtlcSpendSuccess(
+				signer, desc, tx, nil, nil,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			return &Script{
+				Witness: witness,
+			}, nil
+
+		case TaprootHtlcSecondLevelRevoke:
```

Comment:
Still trying to understand the full picture, but it seems `MakeHtlcSecondLevelTimeoutAnchorInput` has some overlap with the `WitnessGenerator` here, can we instead use `WitnessGenerator` inside `MakeHtlcSecondLevelTimeoutAnchorInput`, similar to how it's used in `(bi *BaseInput) CraftInputScript`?

Reply:
Not sure what you mean, this is a revoke path, `MakeHtlcSecondLevelTimeoutAnchorInput` is for normal anchors. 

Might be useful historical context here that: the htlc/contract structs in `input/input.go` came after the witness generation abstraction here.

Reply:
nvm github messed up the comment line🤦🏻